### PR TITLE
Process gitignore-style patterns in `files` property of `package.json`

### LINKED ISSUE #79: Erroneous error when index.d.ts is included in the files list as "/index.d.ts"
**tsd**: 0.13.1 | **Node.js**: v14.9.0 | **OS**: Linux (Arch)

I have a `files` array which includes `"/index.d.ts"`, e.g.:

```json
{
    "files": [
        "dist/index.esm.js",
        "dist/index.js",
        "/index.d.ts"
    ]
}
```

The leading slash is needed to include the `index.d.ts` file in the package's root directory and to exclude other files of the same name, e.g. in the `dist` directory (an unqualified `"index.d.ts"` pattern matches *all* files with that name).

tsd is not picking this up and is failing with an error:

> TypeScript type definition index.d.ts is not part of the files list.

The error goes away if the forward slash is removed.

### PR DESCRIPTION
**Pull request by BendingBender**
Fixes #79

### CODE REVIEW

File: source/lib/rules/files-property.ts [hunk 1]
```diff
@@ -36,3 +39,29 @@ export default (context: Context): Diagnostic[] => {
 		}
 	];
 };
+
+function processGitIgnoreStylePatterns(patterns: string[]): string[] {
+	const processedPatterns = patterns.reduce((includedPatterns, pattern) => {
```
Comment (sindresorhus) [inline source/lib/rules/files-property.ts:44]: I would prefer to not use `.reduce`. It makes code hard to read. Usually, you can solve it with a `.filter`, `.map`, or for-of instead.

File: source/lib/rules/files-property.ts [hunk 2]
```diff
@@ -36,3 +39,29 @@ export default (context: Context): Diagnostic[] => {
...
+		const exclusionMarkers = pattern.match(/^!+/);
```
Comment (sindresorhus) [inline source/lib/rules/files-property.ts:45]: I would prefer using `.filter` and `.startsWith` here.

File: source/lib/rules/files-property.ts [hunk 3]
```diff
@@ -36,3 +39,29 @@ export default (context: Context): Diagnostic[] => {
 		}
 	];
 };
+
+function processGitIgnoreStylePatterns(patterns: string[]): string[] {
```
Comment (sindresorhus) [inline source/lib/rules/files-property.ts:43]:
```suggestion
function processGitIgnoreStylePatterns(patterns: readonly string[]): string[] {
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/lib/rules/files-property.ts (hunk 3 above)
Comment (BendingBender) [reply source/lib/rules/files-property.ts]: done

File: source/lib/rules/files-property.ts (hunk 2 above)
Comment (BendingBender) [reply source/lib/rules/files-property.ts]: `.startsWith()` won't cut it here, we don't know how many markers there are...

File: source/lib/rules/files-property.ts (hunk 1 above)
Comment (BendingBender) [reply source/lib/rules/files-property.ts]: done

**Pull request merged**